### PR TITLE
WIP Parameter names are normalized to not contain colons

### PR DESCRIPTION
**Pull request by mrclay**
Fetched array keys do not contain colons.
For BC, params/types can be still be accessed via names with colons.
Invalid names result in QueryException.

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 1]
```diff
@@ -882,4 +882,55 @@ public function testJoinWithNonUniqueAliasThrowsException()
...
+        $this->assertSame('foo', $qb->getParameter('foo'));
+        $this->assertSame('foo', $qb->getParameter(':foo'));
+
+        $qb->setParameter(':bar', 2, \PDO::PARAM_INT);
+        $this->assertSame(2, $qb->getParameter('bar'));
+        $this->assertSame(2, $qb->getParameter(':bar'));
+
+        $this->assertSame(array(
+            'foo' => 'foo',
+            'bar' => 2,
```
Comment (mrclay) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:900]: The big question is whether this is BC. If user code _expects_ a `:bar` key, it will fail. execute() handles it of course.

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 2]
```diff
@@ -882,4 +882,55 @@ public function testJoinWithNonUniqueAliasThrowsException()
...
+    public function testRejectsInvalidParamNames()
+    {
+        $qb = new QueryBuilder($this->conn);
+
+        $this->setExpectedException(
+            'Doctrine\DBAL\Query\QueryException',
+            "The given name '1a' is invalid."
+        );
+
+        $qb->setParameter('1a', 'one a');
```
Comment (mrclay) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: The plan here is to use PDO's validation strategy http://stackoverflow.com/a/5810058/3779 but allow string integers like "1" to pass through.

File: .gitignore [hunk 3]
```diff
@@ -4,5 +4,6 @@ reports/
 dist/
 download/
 vendor/
+.idea/
```
Comment (Ocramius) [inline .gitignore]: This should be in your global `.gitignore` instead

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 4]
```diff
@@ -1329,6 +1348,36 @@ private function getSQLForJoins($fromAlias, array &$knownAliases)
...
+     * @throws QueryException
+     */
+    private function normalizeKey($key)
+    {
+        if (is_int($key)) {
+            return $key;
+        }
+
+        // int as string
+        if (preg_match('~^(0|[1-9]\\d*)\\z~', $key)) {
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: Can be simplified to `$key === (string) (int) $key`

File: lib/Doctrine/DBAL/Query/QueryException.php [hunk 5]
```diff
@@ -51,4 +51,14 @@ static public function nonUniqueAlias($alias, $registeredAliases)
             "in FROM and JOIN clause table. The currently registered " .
             "aliases are: " . implode(", ", $registeredAliases) . ".");
     }
+
+    /**
+     * @param string $name
+     *
+     * @return \Doctrine\DBAL\Query\QueryException
+     */
+    static public function invalidParamName($name)
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Query/QueryException.php:60]: don't abbreviate `Parameter`

File: lib/Doctrine/DBAL/Query/QueryException.php [hunk 6]
```diff
@@ -51,4 +51,14 @@ static public function nonUniqueAlias($alias, $registeredAliases)
             "in FROM and JOIN clause table. The currently registered " .
             "aliases are: " . implode(", ", $registeredAliases) . ".");
     }
+
+    /**
+     * @param string $name
+     *
+     * @return \Doctrine\DBAL\Query\QueryException
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Query/QueryException.php]: `@return self` is also sufficient

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php (hunk 1 above)
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:900]: Getting parameters is a rare use-case, but we can document the breakage here. This one seems to be the only affected endpoint

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 7]
```diff
@@ -882,4 +882,77 @@ public function testJoinWithNonUniqueAliasThrowsException()
...
+
+        $qb->setParameter('foo', 'foo');
+        $this->assertSame('foo', $qb->getParameter('foo'));
+        $this->assertSame('foo', $qb->getParameter(':foo'));
+
+        $qb->setParameter(':bar', 2, \PDO::PARAM_INT);
+        $this->assertSame(2, $qb->getParameter('bar'));
+        $this->assertSame(2, $qb->getParameter(':bar'));
+
+        $this->assertSame(array(
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: You can use short arrays: we dropped 5.3 support ages ago :-)

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 8]
```diff
@@ -882,4 +882,77 @@ public function testJoinWithNonUniqueAliasThrowsException()
...
+            'foo' => 'foo',
+            'bar' => 2,
+        ), $qb->getParameters());
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: Note that this formatting is not PSR-2 compliant

File: lib/Doctrine/DBAL/Query/QueryException.php (hunk 6 above)
Comment (mrclay) [inline lib/Doctrine/DBAL/Query/QueryException.php]: Mind if I send another PR to change all the fluent methods to use `self`?

Comment (Ocramius) [inline lib/Doctrine/DBAL/Query/QueryException.php]: Not worth the effort, imo, but appreciated :-)

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php (hunk 1 above)
Comment (mrclay) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:900]: FYI I'm already normalizing in my subclass so this isn't a high priority for me. If you'd rather me just update the docs accordingly to avoid BC changes I can do that instead.

Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:900]: > If you'd rather me just update the docs accordingly to avoid BC changes I can do that instead.

Updating [`UPGRADE.md`](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/UPGRADE.md) is sufficient

File: UPGRADE.md [hunk 9]
```diff
@@ -1,3 +1,15 @@
+## MINOR BC BREAK: Doctrine\DBAL\Query\QueryBuilder
```
Comment (mrclay) [inline UPGRADE.md]: What version do you think this will be in? I figured I'd just leave this out for now.

Comment (Ocramius) [inline UPGRADE.md]: @mrclay 2.6.0

**Comment by mrclay** [comment]
Technically PDO should accept parameters not only starting with digits, but containing only digits, like `:0` or `:00`. I kinda don't know what to do with this fact. Such keys would probably cause problems in DBAL's execution strategy, but a key like `1a` shouldn't cause a problem (and might be used with this class today).

Strawman: If given a key containing only digits, assume it's positional (like this PR). If that key is prepended with `:` or if PHP will interpret the string as an integer key (e.g. `1e2`), throw.

Positional keys: `1`, `"1"`
Named keys: `"foo"`, `"9lives"`
Rejected: `":1"`, `"1e9"`
